core/drivers/fsl_sec/hw_key_blob.c

```diff
@@ -217,20 +224,57 @@ TEE_Result get_hw_unq_key_blob_hw(uint8_t *hw_key, int size)
 TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
 {
 	TEE_Result res;
+	uint8_t temp_buffer[KEY_BLOB_SIZE];
 
 	res = init_and_roll_forward_master_key();
 	if (res != TEE_SUCCESS) {
 		EMSG("Failed to init normal priblob otpmk identity from CAAM");
 		return res;
 	}
 
-	res = get_hw_unq_key_blob_hw(&hwkey->data[0], HW_UNIQUE_KEY_LENGTH);
+	// Use the first 16 bytes for the HUK, the last 16 for the die ID.
+	res = get_hw_unq_key_blob_hw(temp_buffer, KEY_BLOB_SIZE);
+	memcpy(&hwkey->data[0], temp_buffer, HW_UNIQUE_KEY_LENGTH);
 
 	if (res != TEE_SUCCESS) {
 		EMSG("Hardware Unique Key Failed");
 	} else {
 		DMSG("Hardware Unique Key Retrieved:");
-		DHEXDUMP(&hwkey->data[0], HW_UNIQUE_KEY_LENGTH);
+	}
+
+	return res;
+}
+
+int tee_otp_get_die_id(uint8_t *buffer, size_t len)
+{
+	TEE_Result res;
+	uint8_t temp_buffer[KEY_BLOB_SIZE];
+	uint8_t *remaining_buffer = &(temp_buffer[HW_UNIQUE_KEY_LENGTH]);
+	uint32_t remaining_size = KEY_BLOB_SIZE - HW_UNIQUE_KEY_LENGTH;
+
+	// Make sure there is some data left available after taking the HUK.
+	COMPILE_TIME_ASSERT(KEY_BLOB_SIZE > HW_UNIQUE_KEY_LENGTH);
+
+	if (len > TEE_SHA256_HASH_SIZE) {
```

Comment:
Bad error check, should be if (len < TEE_SHA256_HASH_SIZE) ?

Reply:
This API is very odd, you can request as much data as you want from it but there is a fixed amount of data to return (at least when reading from fuses). OP-TEE maintainers feel SHA256 is the largest this value should ever be (from issue I opened a while ago), so I'm following that. Requesting more than the available data is an error in this case.

tee_hash_createdigest() will truncate the hash to fit the buffer length provided (and fail if the buffer is longer)